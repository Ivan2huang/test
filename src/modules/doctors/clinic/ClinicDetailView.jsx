/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import * as PropTypes from 'prop-types';
import {
  Box,
  Card,
  Divider,
  Hidden,
  Link,
  withStyles,
} from '@material-ui/core';
import { injectIntl } from 'react-intl';

import Typography from '../../../uiComponents/Typography';
import NotificationBox from '../../../uiComponents/NotificationBox';
import { formatMessage, sentenceCase } from '../../../helpers/helpers';
import WarningLg from '../../../icons/WarningLg';
import { trackingAnchor } from '../../../helpers/firebase';
import { CATEGORIES } from '../../../constants/analytics';

const Styles = theme => ({
  itemContainer: {
    background: theme.white,
    boxShadow: '0px 2px 6px rgba(0, 0, 0, 0.05)',
    marginBottom: theme.spacingX(20),
  },
  teleTag: {
    textDecoration: 'none',
    color: theme.mediumEmphasis,
    '&:not(:last-child)::after': {
      content: '", "',
    },
  },
  terminatedItemContainer: {
    marginBottom: theme.spacingX(2),
  },
});

const renderTerminationLabelOfClinicDetail = (intl, classes) => {
  return (
    <NotificationBox className={classes.terminatedItemContainer}>
      <Box pr={2}>
        <WarningLg />
      </Box>
      <Typography type="style-6">
        {formatMessage(
          intl,
          'clinic.detailView.terminationInfo',
          `Terminated. This clinic is no longer part of the panel clinic list and may not be covered.`,
        )}
      </Typography>
    </NotificationBox>
  );
};

const renderAddressSection = (
  classes,
  name,
  consultationType,
  contactNumbers,
  address,
) => (
  <Box p={6}>
    <Box pb={1} component={Typography} type="style-4">
      {name}
    </Box>
    <Box pb={1} component={Typography} type="style-6">
      {sentenceCase(consultationType)}
    </Box>
    <Box pb={1} component={Typography} type="style-6">
      {contactNumbers
        .filter(number => number && number !== '')
        .map(number => (
          <a
            key={number}
            href={`tel:${number}`}
            onClick={trackingAnchor}
            className={classes.teleTag}
            data-category={CATEGORIES.PANEL_CLINIC_DETAIL}
            data-action="Call clinic from clinic details"
          >
            {number}
          </a>
        ))}
    </Box>
    <Box component={Typography} type="style-6">
      {address}
    </Box>
  </Box>
);

const renderAvailabilitySection = (intl, appointmentType) =>
  appointmentType && (
    <>
      <Divider />
      <Box p={6}>
        <Box pb={1} component={Typography} type="style-6">
          {formatMessage(
            intl,
            'clinic.detailView.availability',
            'Availability',
          )}
        </Box>
        <Typography type="style-5">{appointmentType}</Typography>
      </Box>
    </>
  );

const renderConsultationTime = (days, id) => (
  <Box mb={4} key={`${id}-consultationDay`}>
    <Box pb={1} component={Typography} type="style-6">
      {Object.keys(days)[0]}
    </Box>
    {Object.values(days)[0].map((timeSlot, index) => (
      // eslint-disable-next-line react/no-array-index-key
      <Typography key={`${index}-consultationSlot`} type="style-5">
        {timeSlot}
      </Typography>
    ))}
  </Box>
);

const renderConsultationTimingSection = (intl, consultationTimings, peakHour) =>
  !(consultationTimings || peakHour) ? (
    ''
  ) : (
    <>
      <Divider />
      <Box px={6} pt={6} pb={2}>
        {consultationTimings &&
          consultationTimings.map((days, index) =>
            renderConsultationTime(days, index),
          )}
        {peakHour && (
          <Box mb={4}>
            <Box pb={1} component={Typography} type="style-6">
              {formatMessage(intl, 'clinic.detailView.peakHours', 'Peak hours')}
            </Box>
            <Typography type="style-5">{peakHour}</Typography>
          </Box>
        )}
      </Box>
    </>
  );

const renderLanguageSection = (intl, language) =>
  language && (
    <>
      <Divider />
      <Box p={6}>
        <Box pb={1} component={Typography} type="style-6">
          {formatMessage(
            intl,
            'clinic.detailView.languageSpoken',
            'Language(s) spoken',
          )}
        </Box>
        <Typography type="style-5">{language}</Typography>
      </Box>
    </>
  );

const renderClinicDetails = (intl, classes, clinic) => {
  return (
    <>
      {renderAddressSection(
        classes,
        clinic.name,
        clinic.consultationType,
        [clinic.contactNumber1, clinic.contactNumber2, clinic.contactNumber3],
        clinic.address,
        clinic.terminationDate,
        intl,
      )}
      {renderAvailabilitySection(intl, clinic.appointmentType)}
      {renderConsultationTimingSection(
        intl,
        clinic.consultationTimings,
        clinic.peakHour,
      )}
      {renderLanguageSection(intl, clinic.language)}
    </>
  );
};

const ClinicDetailView = ({
  intl,
  classes,
  clinic,
  showClinicDetail,
  backToResults,
}) => {
  if (!clinic.id) return <></>;
  return (
    <>
      <Box
        component={Hidden}
        implementation="css"
        xsUp={!showClinicDetail}
        flex={1}
        mx={8}
      >
        <Box py={6} alignItems="flex-start">
          <Link
            data-testid="btn-back-to-results"
            underline="none"
            component="button"
            onClick={backToResults}
          >
            <Typography type="style-6" color="hyperlink">
              {formatMessage(
                intl,
                'clinic.detailView.clinicDetails',
                'Back to results',
              )}
            </Typography>
          </Link>
        </Box>

        {!!clinic.terminationDate &&
          renderTerminationLabelOfClinicDetail(intl, classes)}

        <Box component={Card} classes={{ root: classes.itemContainer }}>
          {renderClinicDetails(intl, classes, clinic)}
        </Box>
      </Box>
    </>
  );
};

ClinicDetailView.propTypes = {
  intl: PropTypes.shape({}).isRequired,
  classes: PropTypes.exact({
    itemContainer: PropTypes.string.isRequired,
    teleTag: PropTypes.string.isRequired,
    terminatedItemContainer: PropTypes.string.isRequired,
  }).isRequired,
  clinic: PropTypes.shape({}),
  showClinicDetail: PropTypes.bool.isRequired,
  backToResults: PropTypes.func.isRequired,
};

ClinicDetailView.defaultProps = {
  clinic: {},
};

export default injectIntl(withStyles(Styles)(ClinicDetailView));
