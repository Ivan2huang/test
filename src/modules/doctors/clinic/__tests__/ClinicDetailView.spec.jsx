import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import ClinicDetailView from '../ClinicDetailView';
import withIntl from '../../../../i18n/withIntlProvider';
import withTheme from '../../../../themes/withThemeProvider';

jest.mock('../../../../helpers/helpers', () => ({
  formatMessage: jest.fn((intl, id, defaultMessage) => defaultMessage),
  sentenceCase: jest.fn(str => str),
  isEmpty: jest.fn(),
  formatDate: jest.fn().mockReturnValue('2020/03/16'),
}));

describe('Clinic detail view', () => {
  const props = {
    clinic: {
      id: 3580,
      name: 'Dr. Lam Chi Wan Edwin\n',
      contactNumber1: '2157 1111',
      contactNumber2: '',
      contactNumber3: '333 33333',
      address:
        'Room 1311, Central Building, 1-3 Pedder St., Central, Hong Kong',
      consultationType: 'GASTRO ENTEROLOGY & HEPATOLOGY  ',
      specialty: '',
      appointmentType: 'By appointment',
      consultationTimings: [
        {
          'Mon-Fri': ['10:00AM - 1:00PM', '3:00PM - 6:00PM'],
        },
        {
          'Wed / Sat': ['10:00AM - 1:00PM'],
        },
      ],
      peakHour: '10:00AM - 1:00PM',
      language: 'Cantonese and English',
      area: 'Hong Kong Island',
      district: 'Central',
      latitude: 22.2812194,
      longitude: 114.1571841,
      locale: 'en-HK',
      terminationDate: '',
    },
    showClinicDetail: true,
    backToResults: jest.fn(),
  };

  const setUp = (componentProps = props) => {
    const Component = withIntl(withTheme(ClinicDetailView));
    return render(<Component {...componentProps} />);
  };

  it('should match snapshot', () => {
    const { container } = setUp();

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when clinic information is missing', () => {
    const newProps = {
      ...props,
      clinic: { id: 3580, name: 'Dr. Lam Chi Wan Edwin\n' },
    };
    const { container } = setUp(newProps);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when no clinic is selected', () => {
    const newProps = { ...props, clinic: undefined };

    const { container } = setUp(newProps);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when clinic is terminated', () => {
    const newProps = {
      ...props,
      clinic: { ...props.clinic, terminationDate: '12nov2019' },
    };

    const { container } = setUp(newProps);

    expect(container).toMatchSnapshot();
  });

  it('should call backToResult when clicked on back to results link', () => {
    const { getByTestId } = setUp();
    const backToResultsLink = getByTestId('btn-back-to-results');
    fireEvent.click(backToResultsLink);
    expect(props.backToResults).toHaveBeenCalledTimes(1);
  });
});
