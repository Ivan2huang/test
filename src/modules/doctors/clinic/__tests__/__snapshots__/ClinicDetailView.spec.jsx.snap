// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clinic detail view should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-5"
  >
    <div
      class="MuiBox-root MuiBox-root-21"
    >
      <button
        class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiLink-button MuiTypography-colorPrimary"
        data-testid="btn-back-to-results"
      >
        <div
          class="MuiTypography-root makeStyles-root-58 makeStyles-root-59 MuiTypography-body1"
        >
          Back to results
        </div>
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root ClinicDetailView-itemContainer-1 MuiBox-root MuiBox-root-60 ClinicDetailView-itemContainer-1 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-89"
      >
        <div
          class="MuiTypography-root makeStyles-root-91 makeStyles-root-92 MuiBox-root MuiBox-root-90 MuiTypography-body1"
        >
          Dr. Lam Chi Wan Edwin

        </div>
        <div
          class="MuiTypography-root makeStyles-root-94 makeStyles-root-95 MuiBox-root MuiBox-root-93 MuiTypography-body1"
        >
          GASTRO ENTEROLOGY & HEPATOLOGY  
        </div>
        <div
          class="MuiTypography-root makeStyles-root-97 makeStyles-root-98 MuiBox-root MuiBox-root-96 MuiTypography-body1"
        >
          <a
            class="ClinicDetailView-teleTag-2"
            data-action="Call clinic from clinic details"
            data-category="Panel clinic details"
            href="tel:2157 1111"
          >
            2157 1111
          </a>
          <a
            class="ClinicDetailView-teleTag-2"
            data-action="Call clinic from clinic details"
            data-category="Panel clinic details"
            href="tel:333 33333"
          >
            333 33333
          </a>
        </div>
        <div
          class="MuiTypography-root makeStyles-root-100 makeStyles-root-101 MuiBox-root MuiBox-root-99 MuiTypography-body1"
        >
          Room 1311, Central Building, 1-3 Pedder St., Central, Hong Kong
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-108"
      >
        <div
          class="MuiTypography-root makeStyles-root-110 makeStyles-root-111 MuiBox-root MuiBox-root-109 MuiTypography-body1"
        >
          Availability
        </div>
        <div
          class="MuiTypography-root makeStyles-root-112 makeStyles-root-113 MuiTypography-body1"
        >
          By appointment
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-114"
      >
        <div
          class="MuiBox-root MuiBox-root-115"
        >
          <div
            class="MuiTypography-root makeStyles-root-117 makeStyles-root-118 MuiBox-root MuiBox-root-116 MuiTypography-body1"
          >
            Mon-Fri
          </div>
          <div
            class="MuiTypography-root makeStyles-root-119 makeStyles-root-120 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
          <div
            class="MuiTypography-root makeStyles-root-121 makeStyles-root-122 MuiTypography-body1"
          >
            3:00PM - 6:00PM
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-123"
        >
          <div
            class="MuiTypography-root makeStyles-root-125 makeStyles-root-126 MuiBox-root MuiBox-root-124 MuiTypography-body1"
          >
            Wed / Sat
          </div>
          <div
            class="MuiTypography-root makeStyles-root-127 makeStyles-root-128 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-129"
        >
          <div
            class="MuiTypography-root makeStyles-root-131 makeStyles-root-132 MuiBox-root MuiBox-root-130 MuiTypography-body1"
          >
            Peak hours
          </div>
          <div
            class="MuiTypography-root makeStyles-root-133 makeStyles-root-134 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-135"
      >
        <div
          class="MuiTypography-root makeStyles-root-137 makeStyles-root-138 MuiBox-root MuiBox-root-136 MuiTypography-body1"
        >
          Language(s) spoken
        </div>
        <div
          class="MuiTypography-root makeStyles-root-139 makeStyles-root-140 MuiTypography-body1"
        >
          Cantonese and English
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Clinic detail view should match snapshot when clinic information is missing 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-145"
  >
    <div
      class="MuiBox-root MuiBox-root-161"
    >
      <button
        class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiLink-button MuiTypography-colorPrimary"
        data-testid="btn-back-to-results"
      >
        <div
          class="MuiTypography-root makeStyles-root-198 makeStyles-root-199 MuiTypography-body1"
        >
          Back to results
        </div>
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root ClinicDetailView-itemContainer-141 MuiBox-root MuiBox-root-200 ClinicDetailView-itemContainer-141 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-229"
      >
        <div
          class="MuiTypography-root makeStyles-root-231 makeStyles-root-232 MuiBox-root MuiBox-root-230 MuiTypography-body1"
        >
          Dr. Lam Chi Wan Edwin

        </div>
        <div
          class="MuiTypography-root makeStyles-root-234 makeStyles-root-235 MuiBox-root MuiBox-root-233 MuiTypography-body1"
        />
        <div
          class="MuiTypography-root makeStyles-root-237 makeStyles-root-238 MuiBox-root MuiBox-root-236 MuiTypography-body1"
        />
        <div
          class="MuiTypography-root makeStyles-root-240 makeStyles-root-241 MuiBox-root MuiBox-root-239 MuiTypography-body1"
        />
      </div>
      
    </div>
  </div>
</div>
`;

exports[`Clinic detail view should match snapshot when clinic is terminated 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-249"
  >
    <div
      class="MuiBox-root MuiBox-root-265"
    >
      <button
        class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiLink-button MuiTypography-colorPrimary"
        data-testid="btn-back-to-results"
      >
        <div
          class="MuiTypography-root makeStyles-root-302 makeStyles-root-303 MuiTypography-body1"
        >
          Back to results
        </div>
      </button>
    </div>
    <div
      class="MuiBox-root MuiBox-root-307 makeStyles-container-305 makeStyles-container-306 NotificationBox-container-304 ClinicDetailView-terminatedItemContainer-247"
    >
      <div
        class="MuiBox-root MuiBox-root-308"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          style="width: 1.125rem; height: 1.125rem;"
          viewBox="0 0 18 18"
        >
          <path
            d="M9 18C13.9706 18 18 13.9706 18 9C18 4.02944 13.9706 0 9 0C4.02944 0 0 4.02944 0 9C0 13.9706 4.02944 18 9 18Z"
            fill="#FFBB33"
          />
          <path
            d="M9.00098 6.25C9.69133 6.25 10.251 5.69036 10.251 5C10.251 4.30964 9.69133 3.75 9.00098 3.75C8.31062 3.75 7.75098 4.30964 7.75098 5C7.75098 5.69036 8.31062 6.25 9.00098 6.25Z"
            fill="#333333"
          />
          <path
            d="M9.89997 8.25H8.09998V14.25H9.89997V8.25Z"
            fill="#333333"
          />
        </svg>
      </div>
      <div
        class="MuiTypography-root makeStyles-root-318 makeStyles-root-319 MuiTypography-body1"
      >
        Terminated. This clinic is no longer part of the panel clinic list and may not be covered.
      </div>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root ClinicDetailView-itemContainer-245 MuiBox-root MuiBox-root-320 ClinicDetailView-itemContainer-245 MuiPaper-rounded"
    >
      <div
        class="MuiBox-root MuiBox-root-349"
      >
        <div
          class="MuiTypography-root makeStyles-root-351 makeStyles-root-352 MuiBox-root MuiBox-root-350 MuiTypography-body1"
        >
          Dr. Lam Chi Wan Edwin

        </div>
        <div
          class="MuiTypography-root makeStyles-root-354 makeStyles-root-355 MuiBox-root MuiBox-root-353 MuiTypography-body1"
        >
          GASTRO ENTEROLOGY & HEPATOLOGY  
        </div>
        <div
          class="MuiTypography-root makeStyles-root-357 makeStyles-root-358 MuiBox-root MuiBox-root-356 MuiTypography-body1"
        >
          <a
            class="ClinicDetailView-teleTag-246"
            data-action="Call clinic from clinic details"
            data-category="Panel clinic details"
            href="tel:2157 1111"
          >
            2157 1111
          </a>
          <a
            class="ClinicDetailView-teleTag-246"
            data-action="Call clinic from clinic details"
            data-category="Panel clinic details"
            href="tel:333 33333"
          >
            333 33333
          </a>
        </div>
        <div
          class="MuiTypography-root makeStyles-root-360 makeStyles-root-361 MuiBox-root MuiBox-root-359 MuiTypography-body1"
        >
          Room 1311, Central Building, 1-3 Pedder St., Central, Hong Kong
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-368"
      >
        <div
          class="MuiTypography-root makeStyles-root-370 makeStyles-root-371 MuiBox-root MuiBox-root-369 MuiTypography-body1"
        >
          Availability
        </div>
        <div
          class="MuiTypography-root makeStyles-root-372 makeStyles-root-373 MuiTypography-body1"
        >
          By appointment
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-374"
      >
        <div
          class="MuiBox-root MuiBox-root-375"
        >
          <div
            class="MuiTypography-root makeStyles-root-377 makeStyles-root-378 MuiBox-root MuiBox-root-376 MuiTypography-body1"
          >
            Mon-Fri
          </div>
          <div
            class="MuiTypography-root makeStyles-root-379 makeStyles-root-380 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
          <div
            class="MuiTypography-root makeStyles-root-381 makeStyles-root-382 MuiTypography-body1"
          >
            3:00PM - 6:00PM
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-383"
        >
          <div
            class="MuiTypography-root makeStyles-root-385 makeStyles-root-386 MuiBox-root MuiBox-root-384 MuiTypography-body1"
          >
            Wed / Sat
          </div>
          <div
            class="MuiTypography-root makeStyles-root-387 makeStyles-root-388 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-389"
        >
          <div
            class="MuiTypography-root makeStyles-root-391 makeStyles-root-392 MuiBox-root MuiBox-root-390 MuiTypography-body1"
          >
            Peak hours
          </div>
          <div
            class="MuiTypography-root makeStyles-root-393 makeStyles-root-394 MuiTypography-body1"
          >
            10:00AM - 1:00PM
          </div>
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="MuiBox-root MuiBox-root-395"
      >
        <div
          class="MuiTypography-root makeStyles-root-397 makeStyles-root-398 MuiBox-root MuiBox-root-396 MuiTypography-body1"
        >
          Language(s) spoken
        </div>
        <div
          class="MuiTypography-root makeStyles-root-399 makeStyles-root-400 MuiTypography-body1"
        >
          Cantonese and English
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Clinic detail view should match snapshot when no clinic is selected 1`] = `<div />`;
